### DISCUSSION TITLE
Milvus 传文件 然后解析完成向量插向量库 就会崩

### DISCUSSION THREAD

**Discussion by 18734502195** [milvus-io/milvus#36851] [Q&A and General discussion]
一直往向量库的一个collection里面插数据，是1000多页的数据，经过模型解析了，然后向量库就崩掉了
下面为报错日志
1:02:41.795 [task-1] ERROR i.m.c.AbstractMilvusGrpcClient - [logError,2472] - [] - InsertRequest RPC failed! Collection name:rbzw_1362_common
UNAVAILABLE: Network closed for unknown reason
11:03:25.459 [task-1] ERROR i.m.c.AbstractMilvusGrpcClient - [logError,2472] - [] - ReleasePartitionsRequest failed:
syncTimestamp Failed:stack trace: /go/src/github.com/milvus-io/milvus/pkg/tracer/stack_trace.go:51 github.com/milvus-io/milvus/pkg/tracer.StackTrace
/go/src/github.com/milvus-io/milvus/internal/util/grpcclient/client.go:515 github.com/milvus-io/milvus/internal/util/grpcclient.(*ClientBase[...]).Call
/go/src/github.com/milvus-io/milvus/internal/util/grpcclient/client.go:529 github.com/milvus-io/milvus/internal/util/grpcclient.(*ClientBase[...]).ReCall
/go/src/github.com/milvus-io/milvus/internal/distributed/rootcoord/client/client.go:107 github.com/milvus-io/milvus/internal/distributed/rootcoord/client.wrapGrpcCall[...]
/go/src/github.com/milvus-io/milvus/internal/distributed/rootcoord/client/client.go:308 github.com/milvus-io/milvus/internal/distributed/rootcoord/client.(*Client).AllocTimestamp
/go/src/github.com/milvus-io/milvus/internal/proxy/timestamp.go:61 github.com/milvus-io/milvus/internal/proxy.(*timestampAllocator).alloc
/go/src/github.com/milvus-io/milvus/internal/proxy/timestamp.go:86 github.com/milvus-io/milvus/internal/proxy.(*timestampAllocator).AllocOne
/go/src/github.com/milvus-io/milvus/internal/proxy/task_scheduler.go:173 github.com/milvus-io/milvus/internal/proxy.(*baseTaskQueue).Enqueue
/go/src/github.com/milvus-io/milvus/internal/proxy/task_scheduler.go:350 github.com/milvus-io/milvus/internal/proxy.(*ddTaskQueue).Enqueue
/go/src/github.com/milvus-io/milvus/internal/proxy/impl.go:1368 github.com/milvus-io/milvus/internal/proxy.(*Proxy).ReleasePartitions: attempt #0: standalone=8 stage=Abnormal: service not ready: attempt https://github.com/zilliztech/attu/discussions/1: rpc error: code = Canceled desc = grpc: the client connection is closing: attempt https://github.com/zilliztech/attu/pull/2: rpc error: code = Canceled desc = grpc: the client connection is closing: attempt

**Comment by yhmo** [comment]
Standalone? 有可能是OOM
机器的可用内存是多少？
1000多页是多少条数据？维度是？有没有长字符串insert进去？

可以尝试把milvus.yaml里这几项改小，全都改成0.5试试
```
quotaAndLimits:
  limitWriting:
    memProtection:
      # When memory usage > memoryHighWaterLevel, all dml requests would be rejected;
      # When memoryLowWaterLevel < memory usage < memoryHighWaterLevel, reduce the dml rate;
      # When memory usage < memoryLowWaterLevel, no action.
      enabled: true
      dataNodeMemoryLowWaterLevel: 0.85 # (0, 1], memoryLowWaterLevel in DataNodes
      dataNodeMemoryHighWaterLevel: 0.95 # (0, 1], memoryHighWaterLevel in DataNodes
      queryNodeMemoryLowWaterLevel: 0.85 # (0, 1], memoryLowWaterLevel in QueryNodes
      queryNodeMemoryHighWaterLevel: 0.95 # (0, 1], memoryHighWaterLevel in QueryNodes
```

**Comment by 18734502195** [comment]
修改完后报错
InsertRequest failed:
deny to write, reason: memory limit exceeded, req: /milvus.proto.milvus.MilvusService/Insert: force deny

**Reply by xiaofan-luan** [reply]
看起来内存放不下这么多数据。
正常HNSW索引8GB内存可以放150w 768 dim，你可以大概算一下

**Comment by 18734502195** [comment]
我们是循环插入的话 会占很大的内存吗

**Reply by yhmo** [reply]
建表之后先不要调用collection.load()，然后插入数据，这样内存用量会低于某个水平，只有datanode在使用内存。
如果你建好表就调用了collection.load()，那么插入的数据会立刻驻留在querynode的内存里，随着数据增多，内存也会增多。在standalone中，datanode和querynode同时使用过多内存，就有可能oom。

**Comment by 18734502195** [comment]
总共9000次循环 每次循环插入数据之前都调了这个方法   milvusClient.loadCollection(LoadCollectionParam.newBuilder().withCollectionName(collection).build());    插入后调了 milvusClient.flush(FlushParam.newBuilder().addCollectionName(collection).withSyncFlush(false).build()); 这个方法              现在本地测试的时候把这2个方法注释掉了 向量库就正常了 数据能插入进去， 但是因为是循环插入，很慢 想问下有没有批量的方法能优化下

**Reply by yhmo** [reply]
flush()接口会把缓存里的数据强制落盘，每插一点数据就落盘会形成大量的碎片分片，milvus内部会自动对碎片做合并，把多个小的合成大的，这个过程会额外消耗内存，随着合并的结果越来越大，合并操作所需的内存消耗也会增加。你又load了collection，插入的数据都要驻留在内存里，另外还要给这些数据建索引，消耗的内存以数倍计。

我们不建议自己去调用flush，除非你清楚地知道这个接口干了什么。你就只管insert，让milvus自己管理分片和数据落盘。

milvus本质上是个内存数据库，对机器内存有要求，内存太低的机器跑起来够呛。

**Reply by xiaofan-luan** [reply]
根据循环调用的经验，一般是写了bug，第一次插入一条，第二次插入两条，9000次插入了4000w条。
建议先检查代码
